sql/time_bucket_ng.sql

```diff
@@ -9,21 +9,32 @@
 -- and the interface of this function are subjects to change. There could
 -- be bugs, and the implementation doesn't claim to be complete. Use at
 -- your own risk.
+--
+-- This function is IMMUTABLE while similar PostgreSQL function, date_trunc(),
+-- is STABLE. The difference is that date_trunc() implicitly uses the current
+-- TZ to calculate the result. This is not the case with time_bucket_ng(),
+-- because it's going to get timezone explicitly in text format, e.g.:
+--
+--   SELECT time_bucket_ng('3 months', now() :: timestamptz, 'Europe/Berlin');
+--
+-- In other words, there will be separate continuous aggregates for each
+-- timezone. The timestamptz argument is not a problem, since internally it's
+-- represented as a simple UTC timestamp.
 CREATE OR REPLACE FUNCTION timescaledb_experimental.time_bucket_ng(bucket_width INTERVAL, ts DATE) RETURNS DATE
-	AS '@MODULE_PATHNAME@', 'ts_time_bucket_ng' LANGUAGE C STABLE PARALLEL SAFE STRICT;
+	AS '@MODULE_PATHNAME@', 'ts_time_bucket_ng' LANGUAGE C IMMUTABLE PARALLEL SAFE STRICT;
```

Comment:
The function isn't taking an explicit time zone as argument right now, so why make it IMMUTABLE before that functionality is available. 

What is the behavior in the meantime? Do we hard-code a time zone as if the function was always called with that explicit time zone? Or, will the behavior not reflect the volatility class?

FWIW, it makes more sense to me to first add time zone argument support, then change the volatility.

Reply:
@erimatnor Thanks for the good question!

The reason is that I'm already facing problems because of the fact that the procedure is STABLE. If you try to use it in continuous aggregates you get "only immutable functions supported in continuous aggregate view" error and this is not easy to bypass because functions you could use to do it are declared as static in PostgreSQL code.

Reply:
From https://www.postgresql.org/docs/current/xfunc-volatility.html
```
An IMMUTABLE function cannot modify the database and is guaranteed to return the same results given the same arguments forever.
```
Any function dealing with timestamptz and timezones cannot be IMMUTABLE since rules about timezones change all the time.

Reply:
Thanks, this is actually not a bad point. But the rules change only in the future. If now the time is 2021-06-22 11:39 MSK it will never change 🤷‍♂️ I don't think that any of our users going to work with timestamptz's from the future, but we can add a corresponding check and throw an error if they do. Then we are fine to use IMMUTABLE, right?

Reply:
I realized that we are not, by the definition. Have to think a little about it.